src/envoy/extensions/wasm/wasm.cc

```diff
@@ -115,22 +134,94 @@ WasmHandleExtensionCloneFactory IstioWasmExtension::wasmCloneFactory() {
   };
 }
 
+static std::string statsKey(const PluginSharedPtr& plugin) {
+  auto sep = std::string("\t");
+  return plugin->name_ + sep + plugin->runtime_;
+}
+
 void IstioWasmExtension::onEvent(WasmEvent event,
                                  const PluginSharedPtr& plugin) {
   EnvoyWasm::onEvent(event, plugin);
+  auto key = statsKey(plugin);
+  auto& stats = config_stats_.at(key);
+  switch (event) {
+    case EnvoyWasm::WasmEvent::Ok:
+    case EnvoyWasm::WasmEvent::RemoteLoadCacheHit:
+      break;
+    case EnvoyWasm::WasmEvent::RemoteLoadCacheNegativeHit:
+      stats->permanent_read_error_->inc();
+      break;
+    case EnvoyWasm::WasmEvent::RemoteLoadCacheMiss:
+      stats->eventually_consistent_read_->inc();
+      break;
+    case EnvoyWasm::WasmEvent::RemoteLoadCacheFetchSuccess:
+      break;
+    case EnvoyWasm::WasmEvent::RemoteLoadCacheFetchFailure:
+      stats->permanent_read_error_->inc();
+      break;
+    case EnvoyWasm::WasmEvent::UnableToCreateVM:
+    case EnvoyWasm::WasmEvent::UnableToCloneVM:
+    case EnvoyWasm::WasmEvent::MissingFunction:
```

Comment:
We should add metrics for these too later.

Reply:
If you tell me what you want to call them and how they group up I can do it now.

Reply:
Does UnableToCloneVM happen due to a memory error, do we NACK in all these cases?

Reply:
This error can occur either during initial configuration when we create a canary VM which we use to test the configuration in which case will do NACK the configuration.  It can also occur asynchronously when we are doing the per-Worker initialization. At this point we can't NACK so we don't.  Instead we fail closed or open depending on configuration.